### PR TITLE
Fix PyQt installation

### PR DESCRIPTION
**Pull request by adamjstewart**
Closes #12353 

@Bidibulke @Sinan81 can you test this? It fixes both `py-pyqt4` and `py-pyqt5` and should be generic enough that it can be extended to new SIP packages. I also added additional import tests that should catch the bug before it happens again.

### CODE REVIEW

**Comment by adamjstewart** [comment]
For the record, I wasn't actually able to test `py-pyqt4` because of this error on macOS:
```
/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/qt-4.8.7-iapsr2pnoomqo6ajjcthvfilhhhvxu4y/include/QtCore/qglobal.h:68:10: fatal error: 'algorithm' file not found
#include <algorithm>
         ^~~~~~~~~~~
1 warning and 1 error generated.
make: *** [qtdetail.o] Error 1
```
If anyone can figure out how to get past this, let me know.

**Comment by adamjstewart** [comment]
Hmm, import tests seem to be failing:
```
==> [2019-08-19-20:10:49.914222] '/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/python-3.7.4-jqlxzxpwzdhmrirodxanluofnce5o3ab/bin/python3.7' '-c' 'import PyQt5'
==> [2019-08-19-20:10:49.969236] '/Users/Adam/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/python-3.7.4-jqlxzxpwzdhmrirodxanluofnce5o3ab/bin/python3.7' '-c' 'import PyQt5.QtCore'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'PyQt5.sip'
```

**Comment by adamjstewart** [comment]
Even with `py-sip` as a build/run dependency, the import tests still crash. However, when I set my `PYTHONPATH` locally and try to import the module, I get a completely different error:
```
>>> import PyQt5.QtCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'PyQt5.QtCore'
```
Not really sure what's going on here. Perhaps there is some kind of namespace collision since both `py-sip` and `py-pyqt5` put their Python modules in `lib/python3.7/site-packages/PyQt5`? @lee218llnl @citibeth is this something that can be fixed by adding a `setuptools` dependency?

**Comment by Sinan81** [comment]
I tried it on Qt4 (without +phonon+dbus options as build fails in that case). It seems, the situation is similar but interestingly enough if one attemps to import `PyQt4.QtCore` twice, it works!
```
$ source <( spack module tcl loads --dependencies py-pyqt4 )
sinan@ws-067 /mnt/local/spack (try_adamjstwert_pyqt_sip_fix)
$ module list
Currently Loaded Modulefiles:
  1) bzip2-1.0.8-gcc-8.2.0-vyrgw5d                19) libpng-1.6.34-gcc-8.2.0-nu6yixy              37) xextproto-7.3.0-gcc-8.2.0-lk677zo
  2) libbsd-0.9.1-gcc-8.2.0-64drktg               20) freetype-2.9.1-gcc-8.2.0-qqcute2             38) libx11-1.6.7-gcc-8.2.0-fde7vhi
  3) expat-2.2.5-gcc-8.2.0-yueu76y                21) fontconfig-2.12.3-gcc-8.2.0-a6jxx5v          39) libxext-1.3.3-gcc-8.2.0-m5gjspr
  4) ncurses-6.1-gcc-8.2.0-i655y27                22) pcre-8.42-gcc-8.2.0-zwggkj7                  40) util-macros-1.19.1-gcc-8.2.0-4smpt37
  5) readline-7.0-gcc-8.2.0-wcnxwmb               23) perl-5.26.2-gcc-8.2.0-ndoqgto                41) xkbdata-1.0.1-gcc-8.2.0-6zxoqhg
  6) gdbm-1.18.1-gcc-8.2.0-2ggaajb                24) glib-2.56.3-gcc-8.2.0-gvfobiw                42) libxkbcommon-0.8.2-gcc-8.2.0-pqpyro5
  7) libiconv-1.15-gcc-8.2.0-q7w32zb              25) icu4c-60.1-gcc-8.2.0-yaq7rkn                 43) xcb-util-0.4.0-gcc-8.2.0-irdjmm5
  8) xz-5.2.4-gcc-8.2.0-fvj6toh                   26) inputproto-2.3.2-gcc-8.2.0-zdtqu6u           44) xcb-util-image-0.4.0-gcc-8.2.0-3zvzcz2
  9) zlib-1.2.11-gcc-8.2.0-6wdekzy                27) libjpeg-turbo-2.0.2-gcc-8.2.0-gxsucwx        45) xcb-util-keysyms-0.4.0-gcc-8.2.0-ljwbtlh
 10) libxml2-2.9.9-gcc-8.2.0-ymjgx5m              28) libtiff-4.0.10-gcc-8.2.0-kpr7evk             46) xcb-util-renderutil-0.3.9-gcc-8.2.0-btbc4a6
 11) tar-1.31-gcc-8.2.0-t2cmxty                   29) lcms-2.9-gcc-8.2.0-gqfg3qh                   47) xcb-util-wm-0.4.1-gcc-8.2.0-ikqxp6j
 12) gettext-0.19.8.1-gcc-8.2.0-jtb3ioo           30) libmng-2.0.3-gcc-8.2.0-b2ckdgy               48) qt-4.8.6-gcc-8.2.0-moyj5eo
 13) libffi-3.2.1-gcc-8.2.0-2ssqtqu               31) kbproto-1.0.7-gcc-8.2.0-2vjelnv              49) py-pyqt4-4.12.1-gcc-8.2.0-lckb3rj
 14) openssl-1.0.2r-gcc-8.2.0-tl7ajk4             32) libpthread-stubs-0.4-gcc-8.2.0-bsy6nqs       50) qt-4.8.7-gcc-8.2.0-gsjwzi4
 15) sqlite-3.28.0-gcc-8.2.0-5rujt2k              33) xproto-7.0.31-gcc-8.2.0-3wym7qv              51) py-pyqt4-4.12.1-gcc-8.2.0-in2sqsz
 16) python-2.7.16-gcc-8.2.0-juswpp7              34) libxau-1.0.8-gcc-8.2.0-agyejqp               52) py-sip-4.19.18-gcc-8.2.0-twdfi7k
 17) py-sip-4.19.18-gcc-8.2.0-hfdig27             35) libxdmcp-1.1.2-gcc-8.2.0-foxpnlx             53) py-pyqt4-4.12.3-gcc-8.2.0-erhzuos
 18) font-util-1.3.1-gcc-8.2.0-iq7sr6t            36) libxcb-1.13-gcc-8.2.0-be6phqf
sinan@ws-067 /mnt/local/spack (try_adamjstwert_pyqt_sip_fix)
$ which python
/mnt/local/spack/opt/spack/linux-centos7-x86_64/gcc-8.2.0/python-2.7.16-juswpp7sgu3yu36znxgw3np7dmwjql7e/bin/python
sinan@ws-067 /mnt/local/spack (try_adamjstwert_pyqt_sip_fix)
$ python
Python 2.7.16 (default, Aug  4 2019, 15:54:54)
[GCC 8.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import PyQt4.QtCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named sip
>>> import PyQt4
>>> import PyQt4.QtCore
>>>
sinan@ws-067 /mnt/local/spack (try_adamjstwert_pyqt_sip_fix)
$ python
Python 2.7.16 (default, Aug  4 2019, 15:54:54)
[GCC 8.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import PyQt4
>>> import PyQt4.QtCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named sip
>>> import PyQt4.QtCore
>>>
```

**Comment by adamjstewart** [comment]
@Sinan81 Are you able to build your other packages that depend on PyQt using these changes?

**Comment by Sinan81** [comment]
> Even with `py-sip` as a build/run dependency, the import tests still crash. However, when I set my `PYTHONPATH` locally and try to import the module, I get a completely different error:
>...
> Not really sure what's going on here. ...

FYI, it is mentioned in a related [Guix webpage](https://git.savannah.gnu.org/cgit/guix.git/plain/gnu/packages/qt.scm) that pyqt disregards PYTHONPATH

**Comment by Sinan81** [comment]
Re-tried compiling Qscintilla with it. Getting the usual error:
```
 >> 322    Error: Unable to import PyQt5.QtCore. Make sure PyQt5 is installed.
```
Encouraged by the above mentioned observation, I tried importing QtCore twice which worked and `configure.py` script proceeded to the next step, where another error shows up:
```
sinan@ws-067 /cache/spack-stage/sbulut/qscintilla-2.10.2-gjgjseltppns2eimi2c3qtylf6xztzbf/spack-src/Python
$ python ./configure.py --pyqt=PyQt5                                       
An internal error occured.  Please report all the output from the program,
including the following traceback, to support@riverbankcomputing.com.
Traceback (most recent call last):
  File "./configure.py", line 1867, in <module>                            
    _main(sys.argv, pkg_config)                                               
  File "./configure.py", line 1744, in _main                                  
    target_config.introspect_pyqt(pkg_config)                                
  File "./configure.py", line 874, in introspect_pyqt
    self.pyqt_version_str = QtCore.PYQT_VERSION_STR                           
AttributeError: 'module' object has no attribute 'PYQT_VERSION_STR'
```
Tomorrow, I will try compiling other packages that are known to build before.

**Comment by lee218llnl** [comment]
I have had issues with qt packages in the past. Even with this PR, I run into issues 
[spack-build-out.txt](https://github.com/spack/spack/files/3522023/spack-build-out.txt). I attached my build output for `spack install py-pyqt5`, which fails when installing the base qt package. This might be outside of the scope of this PR, but I figure it worth chiming in.

**Comment by Sinan81** [comment]
@lee218llnl that is a qt4 compilation problem. Qt@4 will compile just fine without phonon and dbus variants. I managed to resolve that particular error (i.e "gstreamer.cpp:42:10: fatal error: gst/gst.h: No such file or directory") by creating gstreamer packages, and explicitly including them in qt similar to how dbus is included in qt package file:
```
        if '+dbus' in self.spec:
            dbus = self.spec['dbus'].prefix
            config_args.append('-dbus-linked')
            config_args.append('-I%s/dbus-1.0/include' % dbus.lib)
            config_args.append('-I%s/dbus-1.0' % dbus.include)
            config_args.append('-L%s' % dbus.lib)
``` 
While compiling qt4, there are a few other similar errors showing up. We should create a separate issue for this i guess.

**Comment by lee218llnl** [comment]
@Sinan81 if I create a new issue, would you mind taking the assignment and try generating a PR?

**Comment by Sinan81** [comment]
Sure. I can look into it.

**Comment by adamjstewart** [comment]
@Bidibulke @Sinan81 so what do we want to do here? Worst comes to worst, we can add `py-sip` as a resource for each `SIPPackage` and build that package's module if things really need to be in the same directory.

**Comment by Sinan81** [comment]
Can you explain what you mean by 'adding py-sip as resource'? 

To be honest, my understanding of this sip-module thing is limited at this moment, so I don't have any useful suggestions.

**Comment by adamjstewart** [comment]
By resources I mean [Spack `resource()`](https://spack.readthedocs.io/en/latest/packaging_guide.html#resources-expanding-extra-tarballs). We could install a second copy of `py-sip` in the same installation prefix as `py-pyqt?` so that they're in the same location.

**Comment by Sinan81** [comment]
@adamjstewart Have you managed to resolve the "ImportError: No module named sip" yet?

**Comment by adamjstewart** [comment]
Nope, I haven't tried the resource thing yet because I haven't had time. Do you have any other ideas?

**Comment by Bidibulke** [comment]
If you need a wrapper of Qt in Python, you can use PySide2 instead of PyQt5.
I'm developing a recipe for this package, but still have some issue with linking with the good C++ lib.
If you are interested, I can share it.

**Comment by Sinan81** [comment]
@adamjstewart I think the resource thing might just work. Today I did a quick hack where I symlinked py-sip_prefix/.../PyQt5/sip.so to py-qt5_prefix/.../PyQt5/ and the import error disappeared and python bindings for Qscintilla  compiled just fine. I think having py-sip installation under py-qt might lead to a similar resolution of the import error.

**Comment by adamjstewart** [comment]
@Sinan81 Thanks, I'll try to put something together by this weekend.

**Comment by Sinan81** [comment]
On a side note, yesterday I noticed that PyQt "*.sip" files are installed under python prefix. Is this a bug?
```
sbulut@ws-067 /mnt/local/spack/opt/spack/linux-centos7-x86_64/gcc-8.2.0/python-2.7.16-57cs32feqysd6l3tugwch4zxgmafqjlu/share/sip/PyQt5 (try_qscintilla_ajs_pyqtsip) 
$ ls
Enginio          QtMacExtras          QtQuick          QtWebKitWidgets
QAxContainer     QtMultimedia         QtQuickWidgets   QtWebSockets
QtAndroidExtras  QtMultimediaWidgets  QtRemoteObjects  QtWidgets
QtBluetooth      QtNetwork            QtSensors        QtWinExtras
QtCore           QtNetworkAuth        QtSerialPort     QtX11Extras
QtDBus           QtNfc                QtSql            QtXml
QtDesigner       QtOpenGL             QtSvg            QtXmlPatterns
QtGui            QtPositioning        QtTest
QtHelp           QtPrintSupport       QtWebChannel
QtLocation       QtQml                QtWebKit
```

**Comment by adamjstewart** [comment]
@Sinan81 yes, that looks like a bug, unless you ran `spack activate` at some point, in which case Spack creates symlinks to the `python` installation prefix.

**Comment by Sinan81** [comment]
I didn't do `spack activate` but just `spack load`
I will double check that I am using your latest commit for pyqt

**Comment by Sinan81** [comment]
According to help page of `configure.py` for PyQt5, default installation directory for sip files is under python prefix
```
sage: python configure.py [opts] [name=value] [name+=value]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  --abi-version=VERSION
                        the SIP ABI version to use (sip5 only)
  -k, --static          build modules as static libraries
  --no-docstrings       disable the generation of docstrings
  -r, --trace           build modules with tracing enabled
  -u, --debug           build modules with debugging symbols
  --qml-debug           enable the QML debugging infrastructure
  -w, --verbose         enable verbose output during configuration
  -c, --concatenate     concatenate each module's C++ source files
  -j N, --concatenate-split=N
                        split the concatenated C++ source files into N pieces
                        [default: 1]

  Configuration:
    --confirm-license   confirm acceptance of the license
    --license-dir=DIR   the license file can be found in DIR [default:
                        /disk/home/sbulut/Downloads/PyQt5_gpl-5.13.0/sip]
    --target-py-version=VERSION
                        the major.minor version of the target Python [default:
                        3.7]
    --link-full-dll     on Windows link against the full Python DLL rather
                        than the limited API DLL
    --sysroot=DIR       DIR is the target system root directory
    --spec=SPEC         pass -spec SPEC to qmake
    --disable=MODULE    disable the specified MODULE [default: checks for all
                        modules will be enabled]
    --disable-feature=FEATURE
                        disable the specified FEATURE
    -e MODULE, --enable=MODULE
                        enable checks for the specified MODULE [default:
                        checks for all modules will be enabled]
    --no-designer-plugin
                        disable the building of the Python plugin for Qt
                        Designer [default: enabled]
    --no-qml-plugin     disable the building of the Python plugin for qmlscene
                        [default: enabled]
    --assume-shared     assume that the Qt libraries have been built as shared
                        libraries [default: check]
    -T, --no-timestamp  suppress timestamps in the header comments of
                        generated code [default: include timestamps]
    --configuration=FILE
                        FILE contains the target configuration
    --protected-is-public
                        enable building with 'protected' redefined as 'public'
                        [default: enabled]
    --protected-not-public
                        disable building with 'protected' redefined as
                        'public'
    --pyuic5-interpreter=FILE
                        the name of the Python interpreter to run the
                        pylupdate5, pyrcc5 and pyuic5 wrappers is FILE
                        [default:
                        /disk/software/lib/hpcpm/spack/opt/spack/linux-centos7
                        -x86_64/gcc-7.4.0/python-3.7.4-mbxbk5xl34brqfl6542u3ra
                        tm456rmkl/bin/python]
    -q FILE, --qmake=FILE
                        the pathname of qmake is FILE [default: search PATH]
    --sip=FILE          the pathname of sip is FILE [default: /usr/bin/sip]
    --sip-incdir=DIR    the directory containing the sip.h header file is DIR
                        [default: None]
    --allow-sip-warnings
                        allow sip to issue non-fatal warning messages
                        [default: warning messages are treated as errors]
    --no-python-dbus    disable the Qt support for the standard Python DBus
                        bindings [default: enabled]
    -s DIR, --dbus=DIR  the directory containing the dbus/dbus-python.h header
                        is DIR [default: supplied by pkg-config]

  Installation:
    -b DIR, --bindir=DIR
                        install pyuic5, pyrcc5 and pylupdate5 in DIR [default:
                        /disk/software/lib/hpcpm/spack/opt/spack/linux-centos7
                        -x86_64/gcc-7.4.0/python-3.7.4-mbxbk5xl34brqfl6542u3ra
                        tm456rmkl/bin]
    -d DIR, --destdir=DIR
                        install the PyQt5 Python package in DIR [default:
                        /disk/software/lib/hpcpm/spack/opt/spack/linux-centos7
                        -x86_64/gcc-7.4.0/python-3.7.4-mbxbk5xl34brqfl6542u3ra
                        tm456rmkl/lib/python3.7/site-packages]
    --designer-plugindir=DIR
                        install the Python plugin for Qt Designer in DIR
                        [default: QT_INSTALL_PLUGINS/designer]
    --qml-plugindir=DIR
                        install the Python plugin for qmlscene in DIR
                        [default: QT_INSTALL_PLUGINS/PyQt5]
    --no-sip-files      disable the installation of the .sip files [default:
                        enabled]
    -v DIR, --sipdir=DIR
                        install the PyQt5 .sip files in DIR [default:
                        /disk/software/lib/hpcpm/spack/opt/spack/linux-centos7
                        -x86_64/gcc-7.4.0/python-3.7.4-mbxbk5xl34brqfl6542u3ra
                        tm456rmkl/share/sip/PyQt5]
    --no-dist-info      do not install the dist-info directory
    --no-stubs          disable the installation of the PEP 484 stub files
                        [default: enabled]
    --stubsdir=DIR      install the PEP 484 stub files in DIR [default:
                        /disk/software/lib/hpcpm/spack/opt/spack/linux-centos7
                        -x86_64/gcc-7.4.0/python-3.7.4-mbxbk5xl34brqfl6542u3ra
                        tm456rmkl/lib/python3.7/site-packages/PyQt5]
    --no-tools          disable the building of pyuic5, pyrcc5 and pylupdate5
                        [default: enabled]

  VendorID support:
    -i, --vendorid      disabled
    -l DIR, --vendorid-incdir=DIR
                        the VendorID header file is installed in DIR [default:
                        ]
    -m DIR, --vendorid-libdir=DIR
                        the VendorID library is installed in DIR [default: ]

  QScintilla support:
    -a, --qsci-api      always install the PyQt API file for QScintilla
                        [default: install only if QScintilla installed]
    --no-qsci-api       do not install the PyQt API file for QScintilla
                        [default: install only if QScintilla installed]
    -n DIR, --qsci-api-destdir=DIR
                        install the PyQt5 API file for QScintilla in DIR
                        [default: QT_INSTALL_DATA/qsci]
~
```

**Comment by adamjstewart** [comment]
> On a side note, yesterday I noticed that PyQt "*.sip" files are installed under python prefix.

I was able to reproduce this. I'll fix it.

File: var/spack/repos/builtin/packages/py-pyqt4/package.py [hunk 1]
```diff
@@ -14,18 +14,30 @@ class PyPyqt4(SIPPackage):
...
+        'PyQt4.QtTest', 'PyQt4.QtWebKit', 'PyQt4.QtXml', 'PyQt4.QtXmlPatterns'
+    ]
 
     version('4.12.3', sha256='a00f5abef240a7b5852b7924fa5fdf5174569525dc076cd368a566619e56d472')
     version('4.11.3', '997c3e443165a89a559e0d96b061bf70',
             url='http://sourceforge.net/projects/pyqt/files/PyQt4/PyQt-4.11.3/PyQt-x11-gpl-4.11.3.tar.gz')
 
     # Supposedly can also be built with Qt 5 compatibility layer
-    depends_on('qt@:4+phonon+dbus')
+    depends_on('qt@:4')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyqt4/package.py]: @Sinan81 you said you got this package working with `qt@:4~phonon~dbus`? Just want to confirm before I remove this requirement.

**Comment by adamjstewart** [comment]
@Sinan81 @Bidibulke okay, this is ready for a final test. I'm able to successfully install `py-pyqt5` on macOS 10.14.6 with Clang 10.0.1 and Python 3.7.4, it passes all of the new import tests, and none of the files are installed to the Python installation prefix. I'm still not able to install `py-pyqt4` due to that `fatal error: 'algorithm' file not found` error I mentioned above, so if you know how to fix that let me know. Otherwise, I think this PR is finally ready to merge!

File: var/spack/repos/builtin/packages/py-pyqt4/package.py (hunk 1 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/py-pyqt4/package.py]: Will do before end of the day. I am pretty sure 'phonon' is optional.

Comment (Sinan81) [reply var/spack/repos/builtin/packages/py-pyqt4/package.py]: I confirm that py-pyqt4 builds with `qt@4.8.7 ~phonon~dbus`
the specific qt installation which it depends on has the following build options
```
$ spack find -lv qt /st2
==> 1 installed package
-- linux-centos7-x86_64 / gcc@8.2.0 -----------------------------
st2p2z5 qt@4.8.7~dbus~examples~framework freetype=spack ~gtk~krellpatch~opengl patches=4f40b7c6230471c19fd1aea45e264a1642c654023d2da1c8063a05ef095d6586,88ec914c9f2026599cce3258006db3fab48138cb9b37f0a7a182d6731fbe2d4d,a4f37427b34d88df6a6adc16e1e6c805087d4f3a7c658e1797450b0912fcbf5d ~phonon+shared+sql+ssl+tools~webkit
```

**Comment by adamjstewart** [comment]
@Sinan81 thanks for reviewing! If you have any PRs that were waiting for this fix and are now ready for review, just ping me and I'll take a look.

**Comment by dirtydiesel46** [comment]
> @Sinan81 @Bidibulke okay, this is ready for a final test. I'm able to successfully install `py-pyqt5` on macOS 10.14.6 with Clang 10.0.1 and Python 3.7.4, it passes all of the new import tests, and none of the files are installed to the Python installation prefix. I'm still not able to install `py-pyqt4` due to that `fatal error: 'algorithm' file not found` error I mentioned above, so if you know how to fix that let me know. Otherwise, I think this PR is finally ready to merge!

Hi @adamjstewart  - did you ever get to fixing this issue with py-pyqt4 ? Having exactly the same issue with getting pyqt4 working on MacOS Catalina 10.15.5.. 

```/usr/local/include/QtCore/qglobal.h:68:10: fatal error: 'algorithm' file not found
#include <algorithm>
     ^~~~~~~~~~~
1 error generated.
make: *** [qtdetail.o] Error 1
Error: Failed to determine the detail of your Qt installation. Try again using
the --verbose flag to see more detail about the problem.```

**Comment by adamjstewart** [comment]
Nope, unfortunately not.

**Pull request merged**